File: tui/src/ws.ts
import { createSignal } from "solid-js"

type WSEventHandler = (event: { type: string; [key: string]: unknown }) => void

export function createWSClient(baseUrl = "ws://localhost:8000") {
  let ws: WebSocket | null = null
  let sessionId = ""
  let reconnectTimer: ReturnType<typeof setTimeout> | null = null
  let handler: WSEventHandler | null = null
  const [connected, setConnected] = createSignal(false)

  function connect(sid: string) {
    disconnect()
    sessionId = sid
    ws = new WebSocket(`${baseUrl}/ws/${sid}`)

    ws.onopen = () => setConnected(true)

    ws.onmessage = (e) => {
      try {
        const data = JSON.parse(e.data)
        if (data.type === "pong") return
        handler?.(data)
      } catch {}
    }

    ws.onclose = () => {
      setConnected(false)
      reconnectTimer = setTimeout(() => connect(sessionId), 3000)
    }

    ws.onerror = () => {
      ws?.close()
    }
  }

  function disconnect() {
    if (reconnectTimer) {
      clearTimeout(reconnectTimer)
      reconnectTimer = null
    }
    if (ws) {
      ws.onclose = null
      ws.close()
      ws = null
    }
    setConnected(false)
  }

  function send(content: string) {
    if (ws?.readyState === WebSocket.OPEN) {
      ws.send(
        JSON.stringify({
          type: "chat.message",
          content,
          message_id: crypto.randomUUID(),
        })
      )
    }
  }

  function interrupt() {
    if (ws?.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({ type: "chat.interrupt" }))
    }
  }

  function setEventHandler(h: WSEventHandler) {
    handler = h
  }

  return { connect, disconnect, send, interrupt, setEventHandler, connected }
}

File: tui/src/connection/ws.py
import asyncio
import json
import uuid
from typing import Callable, Awaitable

import websockets


class WSClient:
    def __init__(self, base_url: str = "ws://localhost:8000"):
        self.base_url = base_url
        self.ws = None
        self.session_id: str = ""
        self.connected: bool = False
        self._on_event: Callable[[dict], Awaitable[None]] | None = None
        self._recv_task: asyncio.Task | None = None

    def set_event_handler(self, handler: Callable[[dict], Awaitable[None]]):
        self._on_event = handler

    async def connect(self, session_id: str) -> bool:
        self.session_id = session_id
        url = f"{self.base_url}/ws/{session_id}"
        try:
            self.ws = await websockets.connect(url, ping_interval=30, ping_timeout=60)
            self.connected = True
            self._recv_task = asyncio.create_task(self._recv_loop())
            return True
        except Exception:
            self.connected = False
            return False

    async def _recv_loop(self):
        try:
            async for raw in self.ws:
                try:
                    event = json.loads(raw)
                except json.JSONDecodeError:
                    continue
                if self._on_event:
                    await self._on_event(event)
        except websockets.ConnectionClosed:
            pass
        except Exception:
            pass
        finally:
            self.connected = False
            if self._on_event:
                await self._on_event({"type": "connection.lost"})

    async def send_message(self, content: str):
        if not self.ws or not self.connected:
            return
        msg = {
            "type": "chat.message",
            "message_id": str(uuid.uuid4()),
            "content": content,
        }
        await self.ws.send(json.dumps(msg))

    async def interrupt(self):
        if not self.ws or not self.connected:
            return
        await self.ws.send(json.dumps({"type": "chat.interrupt"}))

    async def ping(self):
        if not self.ws or not self.connected:
            return
        await self.ws.send(json.dumps({"type": "ping"}))

    async def disconnect(self):
        if self._recv_task:
            self._recv_task.cancel()
            self._recv_task = None
        if self.ws:
            await self.ws.close()
            self.ws = None
        self.connected = False

    async def reconnect(self) -> bool:
        await self.disconnect()
        if self.session_id:
            return await self.connect(self.session_id)
        return False
